hydra-cluster/src/Hydra/Cluster/Faucet.hs

```diff
@@ -112,6 +123,40 @@ seedFromFaucet RunningNode{networkId, nodeSocket} receivingVerificationKey lovel
     Fuel -> TxOutDatumHash markerDatumHash
     Normal -> TxOutDatumNone
 
+-- | Try to submit tx and retry when some caught exception/s take place.
+retryOnExceptions :: (MonadCatch m, MonadDelay m) => Tracer m FaucetLog -> m () -> m ()
+retryOnExceptions tracer action =
+  action
+    `catches` [ Handler $ \(_ :: SubmitTransactionException) -> do
+                  threadDelay 1
+                  retryOnExceptions tracer action
+              , Handler $ \(ex :: IOException) -> do
+                  unless (isResourceExhausted ex) $
+                    throwIO ex
+                  traceWith tracer $
+                    TraceResourceExhaustedHandled $
+                      "Expected exception raised from seedFromFaucet: " <> show ex
+                  threadDelay 1
+                  retryOnExceptions tracer action
+              ]
+ where
+  isResourceExhausted ex = case ioe_type ex of
+    ResourceExhausted -> True
+    _other -> False
+
+-- | Find the utxo for the corresponding verification key
+-- We expect proper utxo to have more 'Lovelace' than the @lovelace@ argument
+findUTxO :: RunningNode -> Lovelace -> VerificationKey PaymentKey -> Bool -> IO UTxO
+findUTxO RunningNode{networkId, nodeSocket} lovelace faucetVk shouldFilter = do
+  utxos <- queryUTxO networkId nodeSocket QueryTip [buildAddress faucetVk networkId]
+  let foundUTxO =
+        if shouldFilter
+          then UTxO.filter (\o -> txOutLovelace o >= lovelace) utxos
+          else utxos
+  when (null foundUTxO) $
+    throwIO $ NotEnoughFunds{utxos, requestedAmount = lovelace}
+  pure foundUTxO
```

Comment:
Should: not compose functions this way.

This is `seedFromFaucet` specific functionality and the `Bool` flag is smelly. Keep this functionality in `seedFromFaucet`, and not change it .. or change it with the enhancement of selecting multiple UTxo to meet the requested lovelace. (This is what this PR should be about).

If you want to reduce boiler plate of the `queryUTxO networkId nodeSocket QueryTip [buildAddress faucetVk networkId]`, that would be a good function to have a separate `queryFaucetUTxO` or so if you want (but not do refactor too early).